### PR TITLE
Add checksum to ScriptStore cache

### PR DESCRIPTION
**Pull request by tudorms**
### Type of Change
- Bug fix (non-breaking change which fixes an issue)

### Why
We're seeing some crashes that may be caused by corrupt script caches.

### What
Add a checksum to avoid loading a corrupt cache.
 ###### Microsoft Reviewers: codeflow:open?pullrequest=https://github.com/microsoft/react-native-windows/pull/11780&drop=dogfoodAlpha

### CODE REVIEW

File: vnext/Shared/Hasher.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+    std::vector<std::uint8_t> GetHashValue();
+    void HashData(const void* pb, size_t cb);
+
+  private:
+    BCRYPT_ALG_HANDLE m_hAlg;
+    BCRYPT_HASH_HANDLE m_hHash;
+    std::vector<std::uint8_t> m_hashObject;
+
+    SHA256Hasher(const SHA256Hasher&);
```
Comment (vmoroz) [inline vnext/Shared/Hasher.h:19]: >SHA256Hasher [](http://example.com/codeflow?start=4&length=12)

nit: it is better to make them public and add '= delete' in the end.

**Review by vmoroz — APPROVED** [review]
:shipit:

File: vnext/Shared/Hasher.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    {
+      // Failed to create hash
+      throw status;
+    }
+  }
+
+  SHA256Hasher::~SHA256Hasher()
+  {
+    BCryptCloseAlgorithmProvider(m_hAlg, 0);
+    BCryptDestroyHash(m_hHash);
```
Comment (vmoroz) [inline vnext/Shared/Hasher.cpp:38]: >BCryptDestroyHash [](http://example.com/codeflow?start=4&length=17)

I wonder about the destruction order: should we destroy the hash before closing algorithm? I understand that the Hash points to the algorithm.

File: vnext/Shared/Hasher.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+#ifndef NT_SUCCESS
+#define NT_SUCCESS(Status) ((NTSTATUS)(Status) >= 0)
+#endif // NT_SUCCESS
+
+namespace Microsoft::ReactNative {
+
+  SHA256Hasher::SHA256Hasher()
+  {
+    NTSTATUS status;
+    if (!NT_SUCCESS(status = BCryptOpenAlgorithmProvider(&m_hAlg, BCRYPT_SHA256_ALGORITHM, NULL, 0)))
```
Comment (vmoroz) [inline vnext/Shared/Hasher.cpp:13]: > if (!NT_SUCCESS [](http://example.com/codeflow?start=3&length=16)

It is a repetitive pattern. I wonder if it is worth adding a macro that takes care of it.

File: vnext/Shared/Hasher.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    {
+      // Failed to load and initialize a CNG provider
+      throw status;
```
Comment (vmoroz) [inline vnext/Shared/Hasher.cpp:16]: > throw status; [](http://example.com/codeflow?start=5&length=14)

I wonder if it is a good approach to throw while our code does not expect exceptions and supposed to return nullptr if we cannot read/verify the byte code.
My understanding is that we cannot do any good with the class constructor. 
Should we have a simple function that does the hashing that returns std::optional?

**Review by vmoroz — APPROVED** [review]

File: vnext/Shared/BaseScriptStoreImpl.cpp [hunk 5]
```diff
@@ -297,6 +317,15 @@ void BasePreparedScriptStoreImpl::persistPreparedScript(
   prefix->runtimeVersion = runtimeMetadata.version;
   prefix->sizeInBytes = preparedScript->size();
 
+  std::optional<std::vector<std::uint8_t>> hashBuffer =
+      Microsoft::ReactNative::GetSHA256Hash(preparedScript->data(), preparedScript->size());
+  if (!hashBuffer) {
+    // Hashing failed.
+    std::terminate();
```
Comment (NikoAri) [inline vnext/Shared/BaseScriptStoreImpl.cpp:324]: Shouldn't we crash here so we get info for any investigation? From my understanding std::terminate doesn't produce Watson crash telemetry

**Pull request merged**